### PR TITLE
Update dskit with cache refactor changes

### PR DESCRIPTION
**Pull request by lamida**
#### What this PR does
Update to dskit with cache refactor changes. Dskit changes (https://github.com/grafana/dskit/pull/255) must be merged before this merged.

#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by pracucci — COMMENTED** [review]
Can you run `make doc reference-help`? I want to see the impact on the auto-generated doc. We need to better understand what's required to do in Mimir add support for Redis.

**Comment by lamida** [comment]
> Can you run `make doc reference-help`? I want to see the impact on the auto-generated doc. We need to better understand what's required to do in Mimir add support for Redis.

There is no change in the reference doc. The block config for redis was added only under admin_client config which is under EnterpriseConfig. Therfore runing `make doc reference-help` didn't generate any changes.

I was trying to add the redis config under `cache.BackendConfig` so that it can be reusable for chunks-cache, results-cache, etc which now only can use memcached. However, I decided to not pursue it further because redis addition was specifically for admin-cache purpose.

File: pkg/storage/tsdb/bucketcache/caching_bucket_test.go
Comment (56quarters) [inline pkg/storage/tsdb/bucketcache/caching_bucket_test.go:214]: We use the `require` and `assert` packages elsewhere to do things like this.
```suggestion
					require.NoError(t, cache.Delete(context.Background(), cachingKeyObjectSubrange(name, i*subrangeSize, (i+1)*subrangeSize))
```

**Review by 56quarters — APPROVED** [review]
LGTM with one small suggestion

**Comment by gotjosh** [comment]
Hi Team! I'm now blocked from progressing any further with https://github.com/grafana/mimir/issues/4082 until this is resolved.

I've enabled notifications to know when we're done, but I wanted to let you all know - in case there's anything you can do to expedite this.

File: pkg/storage/tsdb/bucketcache/caching_bucket_test.go
Comment (lamida) [reply pkg/storage/tsdb/bucketcache/caching_bucket_test.go]: Updated all. I am also not sure how this test become part of the PR. If I am not mistaken due to some lint error after I reabased.

**Review by pracucci — APPROVED** [review]
LGTM! But please merge it only after https://github.com/grafana/dskit/pull/255 is merged and you've updated `go.mod` here.

**Comment by lamida** [comment]
@pracucci or @56quarters thank you for reviewing this PR. 

I have updated this PR dskit from master branch. Could you please help to merge this, since I don't have merge permission.

One small note, there are some vendor changes I need to push which is not related to my work. go mod-check will fail if I don't push it to this PR. Let me know if I should do anything else intead.

**Review by pracucci — APPROVED** [review]

**Pull request merged**
